code/game/objects/effects/effect_system/smoke.dm

```diff
@@ -419,11 +384,119 @@ datum/effect_system/smoke_spread/tactical
+	S.strength = strength
+	if(S.amount)
+		S.spread_smoke()
+
+/datum/effect_system/smoke_spread/xeno/acid
+	smoke_type = /obj/effect/particle_effect/smoke/xeno/burn
 
-/datum/effect_system/smoke_spread/xeno_weaken
-	smoke_type = /obj/effect/particle_effect/smoke/xeno_weak
+/datum/effect_system/smoke_spread/xeno/neuro
+	smoke_type = /obj/effect/particle_effect/smoke/xeno/neuro
+
+/////////////////////////////////////////////
+// Chem smoke
+/////////////////////////////////////////////
+/obj/effect/particle_effect/smoke/chem
+	lifetime = 10
+	var/list/smoked_mobs
+
+/obj/effect/particle_effect/smoke/chem/Destroy()
+	if(length(smoked_mobs) && alpha) //so the whole cloud won't stop working somehow
+		var/obj/effect/particle_effect/smoke/chem/neighbor = pick(current_cloud)
+		neighbor.chemical_effect()
+
+/obj/effect/particle_effect/smoke/chem/copy_stats(obj/effect/particle_effect/smoke/parent)
+	icon = parent.icon
+	return ..()
+
+/obj/effect/particle_effect/smoke/chem/apply_smoke_effect(turf/T)
+	. = ..()
+	reagents.reaction(T, VAPOR, fraction)
+	for(var/obj/O in T)
+		if(O.type == type)
+			continue
+		if(T.intact_tile && O.level == 1) //hidden under the floor
+			continue
+		reagents.reaction(O, VAPOR, fraction)
+
+/obj/effect/particle_effect/smoke/chem/effect_contact(mob/living/carbon/C)
+	reagents.reaction(C, VAPOR, fraction)
+
+/obj/effect/particle_effect/smoke/chem/effect_inhale(mob/living/carbon/C)
+	if(!length(smoked_mobs))
+		addtimer(CALLBACK(src, .proc/chemical_effect), 4)
+	for(var/obj/effect/particle_effect/smoke/chem/S in current_cloud)
+		smoked_mobs += C
+
+/obj/effect/particle_effect/smoke/chem/proc/chemical_effect()
+	for(var/mob/living/carbon/C in smoked_mobs)
+		reagents.reaction(C, INGEST, fraction / length(smoked_mobs))
+		reagents.copy_to(C, reagents.total_volume, fraction / length(smoked_mobs))
+	for(var/obj/effect/particle_effect/smoke/chem/S in current_cloud)
+		S.smoked_mobs.Cut()
+
+/datum/effect_system/smoke_spread/chem
+	var/obj/chemholder
+	smoke_type = /obj/effect/particle_effect/smoke/chem
+
+/datum/effect_system/smoke_spread/chem/New()
+	. = ..()
+	chemholder = new()
+	chemholder.create_reagents(500)
+
+/datum/effect_system/smoke_spread/chem/set_up(datum/reagents/carry = null, radius = 1, loca, smoke_time, silent = FALSE)
+	if(isturf(loca))
+		location = loca
+	else
+		location = get_turf(loca)
+	range = radius
+	if(smoke_time)
+		lifetime = smoke_time
+
+	carry.copy_to(chemholder, carry.total_volume)
+
+	if(!silent)
+		var/contained = ""
+		for(var/reagent in carry.reagent_list)
+			contained += " [reagent] "
+		if(contained)
+			contained = "\[[contained]\]"
+
+		var/where = "[AREACOORD(location)]"
+		if(carry.my_atom.fingerprintslast)
+			var/mob/M = get_mob_by_key(carry.my_atom.fingerprintslast)
+			var/more = ""
+			if(M)
+				more = "[ADMIN_LOOKUPFLW(M)] "
+			message_admins("Smoke: ([ADMIN_VERBOSEJMP(location)])[contained]. Key: [more ? more : carry.my_atom.fingerprintslast].")
```

Comment:
Could you somehow turn the key into `ADMIN_TPMONTY`? Would make muh adminning easier, not a requirement though.